### PR TITLE
714 rules to OAS // MediaTypesRules

### PR DESCRIPTION
**Pull request by ddufourboivin-zalando**
relates to #714 

A first rule for that issue was converted to OAS (OpenAPI Specification) and the new `Context` object.

For the sake of early feedback, I will open one Pull Request per rule I convert.

Highlights of this PR:
- Convert `MediaTypesRule` to use the `Context` object vs. the `Swagger` object.
- Add `getContextFromFixture` in `TestUtil`.
- Add new _reversed AST JSON pointer_ conversions.
  - `VERB/responses/STATUS_CODE/content/MEDIA_TYPE` to `VERB/responses/STATUS_CODE`
  -  `VERB/requestBody/content/MEDIA_TYPE` to `VERB/consumes`
- Adapt the tests in `MediaTypesRuleTest`
- Change the test method names to a common _Kotlin_ practice (more readable test names).

### CODE REVIEW

File: server/src/main/java/de/zalando/zally/util/ast/JsonPointers.java [hunk 1]
```diff
@@ -26,7 +26,14 @@ private JsonPointers() {
...
         createFn(compile("^/components/securitySchemes/(.*)$"), "/securityDefinitions/%s"),
-        createFn(compile("^/paths/(.+/responses/.+)/content/.+/(schema.*)$"), "/paths/%s/%s")
+        createFn(compile("^/paths/(.+/responses/.+)/content/.+/(schema.*)$"), "/paths/%s/%s"),
+
+        // VERB/responses/STATUS_CODE/content/MEDIA_TYPE --> VERB/responses/STATUS_CODE
+        // Could also be VERB/produces but this information is lost from Swagger 2 to OpenAPI 3 conversion.
+        createFn(compile("^/paths/(.+/responses/.+)/content/[^/]*$"), "/paths/%s"),
+
+        // VERB/requestBody/content/MEDIA_TYPE --> VERB/consumes
+        createFn(compile("^/paths/(.*)/requestBody/content/[^/]*$"),"/paths/%s/consumes")
```
Comment (roxspring) [inline server/src/main/java/de/zalando/zally/util/ast/JsonPointers.java]: Worth adding some explicit unit tests to `JsonPointersTests`??

Comment (ddufourboivin-zalando) [reply server/src/main/java/de/zalando/zally/util/ast/JsonPointers.java]: Completely worth!

Comment (ddufourboivin-zalando) [reply server/src/main/java/de/zalando/zally/util/ast/JsonPointers.java]: @roxspring : Just committed new tests for them. Thanks for the tip.

File: server/src/test/java/de/zalando/zally/rule/zalando/MediaTypesRuleTest.kt [hunk 2]
```diff
@@ -52,58 +50,87 @@ class MediaTypesRuleTest {
...
-        assertThat(rule.validate(swagger)).isNull()
+    fun `the SPA API generates no violations`() {
+        val context = getContextFromFixture("api_spa.yaml")!!
+        assertThat(rule.validate(context)).isEmpty()
     }
+
+    private val rule = MediaTypesRule()
+
+    private fun contextWithMediaTypes(vararg pathToMedia: Pair<String, List<String>>): Context =
+        Context(OpenAPI().apply {
```
Comment (tkrop) [inline server/src/test/java/de/zalando/zally/rule/zalando/MediaTypesRuleTest.kt]: I'm not objecting this kind of coding, but this is really heavy piece code - hard to parse with 12 levels of brackets. Can we think of a more verbose writing for input creation?

Comment (ddufourboivin-zalando) [reply server/src/test/java/de/zalando/zally/rule/zalando/MediaTypesRuleTest.kt]: I thought it would more readable this way because it actually _looks_ like the JSON it would be. Let me try a more _flat_ representation of that code let's compare them.

Comment (ddufourboivin-zalando) [reply server/src/test/java/de/zalando/zally/rule/zalando/MediaTypesRuleTest.kt]: Here are some alternative, @tkrop . Let me know which one you like (same for the other reviewer).

### Version 1

```kotlin
private fun contextWithMediaTypes(vararg pathToMedia: Pair<String, List<String>>): Context =
    Context(OpenAPI().apply {
        paths = pathToMedia.fold(Paths()) { paths, (path, types) ->
            paths.addPathItem(path, PathItem().apply {
                get = Operation().apply {
                    responses = ApiResponses().apply {
                        addApiResponse("200", ApiResponse().apply {
                            content = types.fold(Content()) { content, type ->
                                content.addMediaType(type, MediaType())
                            }
                        })
                    }
                }
            })
        }
    })
```

### Version 2

```kotlin
private fun contextWithMediaTypes2(vararg pathToMedia: Pair<String, List<String>>): Context {

    val paths = pathToMedia.fold(Paths()) { p, (path, types) ->

        val content = types.fold(Content()) { c, type ->
            c.addMediaType(type, MediaType())
        }

        val response = ApiResponse()
        response.content = content

        val responses = ApiResponses()
        responses.addApiResponse("200", response)

        val operation = Operation()
        operation.responses = responses

        val pathItem = PathItem()
        pathItem.get(operation)
        p.addPathItem(path, pathItem)
    }

    val openApi = OpenAPI()
    openApi.paths = paths

    return Context(openApi)
}
```

### Version 3

```kotlin
private fun contextWithMediaTypes3(vararg pathToMedia: Pair<String, List<String>>): Context {
    val paths = pathToMedia.fold(Paths()) { p, (path, types) ->
        val content = types.fold(Content()) { c, type ->
            c.addMediaType(type, MediaType())
        }
        val response = ApiResponse().also {
            it.content = content
        }
        val responses = ApiResponses().also {
            it.addApiResponse("200", response)
        }
        val operation = Operation().also {
            it.responses = responses
        }
        val pathItem = PathItem().also {
            it.get(operation)
        }
        p.addPathItem(path, pathItem)
    }
    val openApi = OpenAPI().also {
        it.paths = paths
    }
    return Context(openApi)
}
```

### Version 4

```kotlin
private fun contextWithMediaTypes4(vararg pathToMedia: Pair<String, List<String>>): Context {
    val paths = Paths()
    for ((path, types) in pathToMedia) {
        val content = Content()
        for (type in types) {
            content.addMediaType(type, MediaType())
        }
        val response = ApiResponse().also {
            it.content = content
        }
        val responses = ApiResponses().also {
            it.addApiResponse("200", response)
        }
        val operation = Operation().also {
            it.responses = responses
        }
        val pathItem = PathItem().also {
            it.get(operation)
        }
        paths.addPathItem(path, pathItem)        }
    val openApi = OpenAPI().also {
        it.paths = paths
    }
    return Context(openApi)
}
```

### Version 5

```kotlin
private fun contextWithMediaTypes5(vararg pathToMedia: Pair<String, List<String>>): Context {
    val paths = Paths()

    for ((path, types) in pathToMedia) {
        val content = Content()
        for (type in types) {
            content.addMediaType(type, MediaType())
        }

        val response = ApiResponse()
        response.content = content

        val responses = ApiResponses()
        responses.addApiResponse("200", response)

        val operation = Operation()
        operation.responses = responses

        val pathItem = PathItem()
        pathItem.get(operation)

        paths.addPathItem(path, pathItem)
    }

    val openApi = OpenAPI()
    openApi.paths = paths

    return Context(openApi)
}
```

Comment (roxspring) [reply server/src/test/java/de/zalando/zally/rule/zalando/MediaTypesRuleTest.kt]: Just throwing my EUR 0.02 into the mix... 

Personally, as you can probably tell from tests I've contributed, I generally find that it's clearest to sidestep this entire issue and define small chunks of actual swagger/openapi inline. To me, building up the model programmatically doesn't read very well and I worry that may not be representing the intended textual content.

Comment (ddufourboivin-zalando) [reply server/src/test/java/de/zalando/zally/rule/zalando/MediaTypesRuleTest.kt]: Not sure I am following, @roxspring . The checks are expecting a `Context` object which is itself expecting a complete spec (`Swagger` or `OpenAPI`) to be provided. How could we _define small chunks of actual swagger/openapi inline_?

Comment (ddufourboivin-zalando) [reply server/src/test/java/de/zalando/zally/rule/zalando/MediaTypesRuleTest.kt]: Actually, if you could point me to a specific test that I could inspire myself from, that would be great! I am new to this project and went kind of the way it was before converting the test, just with the new model (`Context` vs `Swagger`).

Comment (tkrop) [reply server/src/test/java/de/zalando/zally/rule/zalando/MediaTypesRuleTest.kt]: @roxspring Thinking about it, I would prefer creating a short `String` representation over any programmatic approach.

Comment (roxspring) [reply server/src/test/java/de/zalando/zally/rule/zalando/MediaTypesRuleTest.kt]: `AvoidTrailingSlashesRuleTest.negativeCase()` would be a nice neat example of what I mean (also showcasing `ViolationAssert` and `ViolationsAssert`).

So applying this to one of your tests I'd probably have written it as follows rather than building a `contextWithMediaTypes()` method at all:

```kotlin
    @Test
    fun `only some of multiple paths without versioning causes violation`() {
        val content = """
            openapi: 3.0.0
            info:
              title: Test API
              version: 1.0.0
            paths:
              /path1:
                get:
                  responses:
                    200:
                      description: 200 described
                      content:
                        application/json: {}
                        application/vnd.api+json: {}
              /path2:
                get:
                  responses:
                    200:
                      description: 200 described
                      content:
                        application/x.zalando.contract+json: {}
              /path3:
                get:
                  responses:
                    200:
                      description: 200 described
                      content:
                        application/x.zalando.contract+json;v=123: {}
            """.trimIndent()
        val context = Context.createOpenApiContext(content)!!

        val result = rule.validate(context)
        assertThat(result).hasSameElementsAs(listOf(
            v("/paths/~1path1/get/responses/200/content/application~1vnd.api+json"),
            v("/paths/~1path2/get/responses/200/content/application~1x.zalando.contract+json")
        ))
    }
```

FWIW I just had a play at a "middle ground" option but I'm not sure it adds to ease of reading or writing:
```kotlin
    @Test
    fun `only some of multiple paths without versioning causes violation`() {
        val content = contentWithMediaTypes(
                "/path1" to listOf("application/json", "application/vnd.api+json"),
                "/path2" to listOf("application/x.zalando.contract+json"),
                "/path3" to listOf("application/x.zalando.contract+json;v=123")
        )

        val context = Context.createOpenApiContext(content)!!

        val result = rule.validate(context)
        assertThat(result).hasSameElementsAs(listOf(
            v("/paths/~1path1/get/responses/200/content/application~1vnd.api+json"),
            v("/paths/~1path2/get/responses/200/content/application~1x.zalando.contract+json")
        ))
    }

    private fun contentWithMediaTypes(vararg pathToMedia: Pair<String, List<String>>): String {
        return """
            openapi: 3.0.0
            info:
              title: Test API
              version: 1.0.0
            paths:
            """.trimIndent() +
                pathToMedia.joinToString(prefix = "\n", separator = "\n") {
                    """
                    ${it.first}:
                      get:
                        responses:
                          200:
                            description: 200 described
                            content:
                    """.trimIndent().prependIndent("  ") +
                            it.second.joinToString(prefix = "\n", separator = "\n") {"            $it: {}"}
                }
    }
```

Comment (ddufourboivin-zalando) [reply server/src/test/java/de/zalando/zally/rule/zalando/MediaTypesRuleTest.kt]: @roxspring : I am personally a huge fan of seing blocks of JSON or YAML in test and see EXACTLY what is being tested. I did that `contentWithMediaType` to stick as close to the previous implementation of the test as possible.

I will change it and commit!

Comment (ddufourboivin-zalando) [reply server/src/test/java/de/zalando/zally/rule/zalando/MediaTypesRuleTest.kt]: @roxspring @tkrop : Please have a look at [this commit](https://github.com/zalando/zally/pull/740/commits/0eef62639b9bd3eaeb31fe05bcc2de9e3f911456) and tell me what you think.

Comment (tkrop) [reply server/src/test/java/de/zalando/zally/rule/zalando/MediaTypesRuleTest.kt]: @ddufourboivin-zalando @roxspring actually I would favor the second approach:

(1) It contains less redundant code,
(2) It hides the test fixture (https://github.com/junit-team/junit4/wiki/test-fixtures), and
(3) It allows to separate different test cases using the same fixture.

**Comment by maxim-tschumak** [comment]
:+1:

**Comment by ddufourboivin-zalando** [comment]
👍
